core/modulesource.go

```diff
@@ -250,6 +251,153 @@ func (src *ModuleSource) Pin() string {
 	}
 }
 
+func (src *ModuleSource) innerEnvFile(ctx context.Context) (*EnvFile, string, error) {
+	// We only allow loading an env file from local modules, for safety
+	if src.Kind != ModuleSourceKindLocal {
+		return nil, "", nil
+	}
+
+	dag, err := CurrentDagqlServer(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// FIXME: .env must be at the root of the module directory
+	// If the user calls dagger from a subdirectory of the module, and that subdirectory contains a more
+	//  specialized .env, that will be ignored. To fix this, we need access to current workdir on the host,
+	// so that we can findup from there.
+	moduleDirPath := path.Join(
+		src.Local.ContextDirectoryPath, // path of the module's git root, on the host
+		src.SourceRootSubpath,          // path of the module directory, relative to its git root
+	)
+	// Check if the env file exists
+	var envFileExists bool
+	if err := dag.Select(ctx, dag.Root(), &envFileExists,
+		dagql.Selector{Field: "host"},
+		dagql.Selector{
+			Field: "directory",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: dagql.String(moduleDirPath)},
+				{Name: "include", Value: dagql.ArrayInput[dagql.String]{".env"}},
+			},
+		},
+		dagql.Selector{
+			Field: "exists",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: dagql.String(".env")},
+			},
+		},
+	); err != nil {
+		return nil, "", fmt.Errorf("failed to check for inner env file in %s: %w",
+			moduleDirPath, err)
+	}
+	if !envFileExists {
+		return nil, "", nil
+	}
+	envFilePath := path.Join(moduleDirPath, ".env")
+	var envFile *EnvFile
+	if err := dag.Select(ctx, dag.Root(), &envFile,
+		dagql.Selector{Field: "host"},
+		dagql.Selector{
+			Field: "file",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: dagql.String(envFilePath)},
+			},
+		},
```

Comment:
could maybe just try accessing it instead of checking if it exists? or is the idea to avoid scary looking failed telemetry?

Reply:
If I try to access it directly, then I have to process "errors that aren't really errors". I know there's a SDK facility for that, but I prefer to use pure engine API out of principle. The principle being: "make the engine smart and the SDKs dumb, it scales better" :)

Reply:
@sipsma the issue is that I need to differentiate between 1) no `.env` in the module (not an error) and 2) an error occurred while processing this query.

I was trying to avoid having to differentiate 1 and 2 in a SDK-specific way (ie. by checking the type of `error`), but I guess performance is a good reason to do it the SDK-specific way.